### ISSUE TITLE
Add degrotate support for mesh nodes

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#6172]
I suggest to allow to be able to rotate the mesh of nodes with the “mesh” drawtype, using `param2`.

I think the meaning of `paramtype2="degrotate"` could be extended for this.  

Currently it works very well for plantlike nodes, but `degrotate` has no effect for mesh nodes.

Implementing my suggestion would simply mean that `degrotate` also works for meshes. Setting `param2` on such a node will rotate the mesh accordingly.

Example:
```
paramtype2="degrotate",
drawtype="mesh",
mesh="example_standing_sign.b3d",
place_paramtype2 = 45,
```

On placement, the mesh (a standing sign) will be rotated by 45°. By changing the `param2` later, the sign can be rotated freely. It's really just the same as `degrotate`, but for meshes.

Use case:
* Standing sign with fine rotation without needing to register lots of nodes

EDIT: Oh, the same for nodeboxes would be nice as well (optional).

**Comment by raymoo** [comment]
I don't think this would work, since param2 is only a byte

**Comment by Wuzzy2** [comment]
You're wrong. This is how `degrotate` is currently defined.
>     paramtype2 == "degrotate"
>     ^ The rotation of this node is stored in param2. Plants are rotated this way.
>       Values range 0 - 179. The value stored in param2 is multiplied by two to
>       get the actual rotation of the node.

**Comment by Wuzzy2** [comment]
Oh, I like to suggest the same for nodeboxes as well.

**Comment by raymoo** [comment]
Oh, okay, though that's slightly different than the way you described it (the existing one multiplies the value by 2 to get the degree value).

**Comment by Ferk** [comment]
Yes, you wouldn't really be able to rotate a sign to 45°, only to 44° or 46°.

**Comment by paramat** [comment]
:-1: I disagree with degrotate for plantlike and would like to remove it. Meshnodes are similar to normal cubic nodes so should be aligned to the world axes just like cubic nodes, fine rotation is out of character for a voxel game.

However i would support a new drawtype 'upright_spritelike' which is a simple vertical flat plane that can be fine-rotated by degrotate, for use as fine-directional signs.

**Comment by paramat** [comment]
Degrotate was added to plantlike because it is very simple to do so, fine-rotating the vertices of a meshnode is complex, intensive and slow to render. Also, we have little dev time.
A fine-rotatable single plane will be simple to code, i have considered this before.

**Comment by Wuzzy2** [comment]
I can live with an error of 1 degree.

**Comment by paramat** [comment]
Do you want to change this to a request for a fine-rotatable single texture? Otherwise i think this should be closed due to being very complex and not much needed.

**Comment by raymoo** [comment]
@paramat I thought this was already implemented for plantlike? And wouldn't it be simply add a rotation to the transform used to draw the mesh?

**Comment by paramat** [comment]
But why? Degrotate was itself controversial because it created nodes that were not aligned to the world axes or diagonal to the axes. This is a voxel game after all. I wish it had not been added but now we have to support it, let's not make the same mistake again. There's no need, even if not very complex to do, and banners and signs can be done with a single degrotate texture (half a plantlike).

**Comment by raymoo** [comment]
Adding degrotate to meshes would not make anything "worse" since meshes can already be not aligned to the world axes. It just changes 180 node definitions and meshes into one.

**Comment by rubenwardy** [comment]
> Degrotate was itself controversial

It's worth noting that you were basically the only one that didn't like it

**Comment by Wuzzy2** [comment]
No, I will not change my request. If it's only a single plane to rotate, this would be a very different thing.

Paramat, please do not oppose features solely on the fact “This does not belong in XYZ game.”. This argument is already flawed because MT is not itself a game. Let subgames decide how religiously they follow the voxel paradigm. :-)
If a node looks ugly as a result of this new feature, this is **entirely** the fault of the subgame author.

Note that with meshes you are already able to break the voxel paradigm by simply hardcoding the rotation into the node. Basically that's what I already did with signs and standing banners in MCL2.

The reason why I'm asking for this feature is because it would greatly simplify the code.

**Comment by paramat** [comment]
> meshes can already be not aligned to the world axes.

They can, but meshnodes are usually not, because aligning to world axes is more suitable.
So it seems this should be done by the mesh creator not by engine code, because it is not really suitable for a voxel game engine.

> Paramat, please do not oppose features solely on the fact “This does not belong in XYZ game.”. This argument is already flawed because MT is not itself a game.

I'm not, MT is a voxel game engine so i'm saying it is unsuitable, or rather, low priority, for the engine.

> Note that with meshes you are already able to break the voxel paradigm by simply hardcoding the rotation into the node

Yes, but that's no reason to add code for a low priority feature that is somewhat unsuitable for a voxel game engine.

> Basically that's what I already did with signs and standing banners in MCL2.

Signs and banners can be done with a single flat plane that displays a different texture for the back.

**Comment by raymoo** [comment]
@paramat They would look better with some thickness though (IMO). Though if you want to implement extruded drawtypes, then that would work too.

**Comment by paramat** [comment]
I agree, but it's a case of justifying the code, work and future maintenance.

**Comment by raymoo** [comment]
@paramat The change would be simple since there is already a function to rotate meshes on the XZ plane: https://github.com/minetest/minetest/blob/c42753338924bb29c61081c9f269772f89bcd808/src/mesh.cpp#L275

It just needs some glue code similar to plantlike to use this when degrotate is set.

**Comment by paramat** [comment]
Ok good, i'm more neutral then.

**Comment by raymoo** [comment]
I took another look and it might be more complex. If I understand the code correctly, facedir on meshes currently caches each rotation of the mesh for performance. Degrotate might have performance issues if used on large meshes, though a sign probably would not have too many vertices. Caching for degrotate could be implemented, but it would be more expensive than facedir since there could be 180 different versions to cache (as opposed to 24 for facedir I think?).

**Comment by paramat** [comment]
24 yes.
So obviously if done it would have to be a subset of special meshes that can do this, so that only those few meshes have 180 cached versions. I still don't think it's worth doing.

**Comment by raymoo** [comment]
@paramat Alternatively it could have no caching, and modders are advised to limit its use to meshes with low vertex count.

Maybe using an entity shaped like the sign would be better, because it doesn't need to write out a new mesh to rotate.

**Comment by paramat** [comment]
The idea of degrotate dividing 360 degrees into 256 orientation is excellent, i wonder if we can change to this, but of course only after consulting modders, i'm not sure how commonly degrotate is used, i suspect it isn't used much.

**Comment by paramat** [comment]
1 year bump for more core dev opinions.

**Comment by numberZero** [comment]
This should be easy to implement.

Reasonable orientation counts:
* 256 = 2⁸ — only 2-fold rotations
* 180 = 2² ⋅ 3² ⋅ 5 — more variants, but perfect diagonals are lacking
* 120 = 2³ ⋅ 3 ⋅ 5 — still good variants, with perfect diagonals (bonus: 1 bit free)
* 24 = 2³ ⋅ 3 — most useful options, and 3 bits free (e.g. for color)

So, I’d prefer 120 normally, and 24 with color.

**Comment by Wuzzy2** [comment]
Please re-open as this is being actively worked on.

**Comment by paramat** [comment]
Agreed thanks.

**Comment by Wuzzy2** [comment]
Fixed by #7840.

**Comment by HybridDog** [comment]
Would it be possible to store different shapes (in this example case they're 240 shapes, each one is a rotation around Z) in the mesh model file and use param2 to select one of the stored shapes?
